File: integration-tests/src/test/java/dev/capylang/test/ResultBindTest.java
package dev.capylang.test;

import capy.lang.Result;
import org.junit.jupiter.api.Test;

import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;

class ResultBindTest {
    @Test
    void usersReturnsTupleWhenAllConstructorsSucceed() {
        var result = ResultBind.users("Ann", "Bob", "Cid");

        assertThat(result).isInstanceOf(Result.Success.class);
        var value = ((Result.Success<List<Object>>) result).value();
        assertThat(value).hasSize(3);
        assertThat(((ResultBind.User) value.get(0)).name()).isEqualTo("Ann");
        assertThat(((ResultBind.User) value.get(1)).name()).isEqualTo("Bob");
        assertThat(((ResultBind.User) value.get(2)).name()).isEqualTo("Cid");
    }

    @Test
    void usersWithRegularLetKeepsFirstError() {
        assertThat(ResultBind.firstErrorMessage("", "Bob")).isEqualTo("You need to pass name");
        assertThat(ResultBind.firstErrorMessage("Ann", "")).isEqualTo("You need to pass name");
    }
}

File: capy/src/e2e-cfun/js/dev/capylang/test/ResultBind.test.js
const test = require('node:test');
const { assert, assertSuccess, generatedModule } = require('./_support');

test('ResultBind', () => {
    const resultBind = generatedModule('dev/capylang/test/ResultBind.js');
    const users = assertSuccess(resultBind.users('Ann', 'Bob', 'Cid'));
    assert.equal(users.length, 3);
    assert.equal(users[0].name, 'Ann');
    assert.equal(users[2].name, 'Cid');
    assert.equal(resultBind.firstErrorMessage('', 'Bob'), 'You need to pass name');
});
